c++/src/kj/async.c++

```diff
@@ -1201,36 +1186,33 @@ XThreadPaf::XThreadPaf()
     : state(WAITING), executor(getCurrentThreadExecutor()) {}
 XThreadPaf::~XThreadPaf() noexcept(false) {}
 
-void XThreadPaf::Disposer::disposeImpl(void* pointer) const {
-  XThreadPaf* obj = reinterpret_cast<XThreadPaf*>(pointer);
+void XThreadPaf::destroy() {
   auto oldState = WAITING;
 
-  if (__atomic_load_n(&obj->state, __ATOMIC_ACQUIRE) == DISPATCHED) {
+  if (__atomic_load_n(&state, __ATOMIC_ACQUIRE) == DISPATCHED) {
     // Common case: Promise was fully fulfilled and dispatched, no need for locking.
-    delete obj;
-  } else if (__atomic_compare_exchange_n(&obj->state, &oldState, CANCELED, false,
+    delete this;
+  } else if (__atomic_compare_exchange_n(&state, &oldState, CANCELED, false,
                                          __ATOMIC_ACQUIRE, __ATOMIC_ACQUIRE)) {
     // State transitioned from WAITING to CANCELED, so now it's the fulfiller's job to destroy the
     // object.
   } else {
     // Whoops, another thread is already in the process of fulfilling this promise. We'll have to
     // wait for it to finish and transition the state to FULFILLED.
-    obj->executor.impl->state.when([&](auto&) {
-      return obj->state == FULFILLED || obj->state == DISPATCHED;
+    executor.impl->state.when([&](auto&) {
+      return state == FULFILLED || state == DISPATCHED;
     }, [&](Executor::Impl::State& exState) {
-      if (obj->state == FULFILLED) {
+      if (state == FULFILLED) {
         // The object is on the queue but was not yet dispatched. Remove it.
-        exState.fulfilled.remove(*obj);
+        exState.fulfilled.remove(*this);
       }
     });
```

Comment:
I might be mistaken, but I could have sworn that there's a compiler that complains if you try to implicitly use `this` in a lambda without explicitly capturing it.